replica/table.cc

```diff
@@ -470,8 +471,7 @@ void compaction_group_sstables_adder::execute() {
         for (const auto& sst : _sstables) {
             _cg._main_set_disk_space_used += sst->bytes_on_disk();
         }
-        // FIXME: the following isn't exception safe.
-        table::add_sstables_to_backlog_tracker(_cg.get_backlog_tracker(), _sstables);
+        _cg.get_backlog_tracker().swap(*_new_backlog_tracker);
```

Comment:
An alternative way to clone/add is to use software transactions. Each mutator returns an object that, when destroyed, undoes the operation.

The caller calls mutators and stores the objects in some vectors. Once an exception happens, everything gets rolled back. If an exception doesn't happen, we call cancel() on all the reverters.

This favors mutations that are smaller than the object being mutated (since we don't need to copy everything).

/cc @tgrabiec

Reply:
It's possible and seems cleaner.
Just bare in mind that we copy-modify-write the sets also to protect concurrent readers that iterate over the set:
See https://github.com/scylladb/scylladb/blob/af65d5a1e8e61565da6ef60ae7a4a8dab675e0f6/sstables/sstable_set.cc#L326

Reply:
I thought we added versioning to make the copy very cheap.

Reply:
Versioning of sstable_set? Not that I'm aware of.

Reply:
> I thought we added versioning to make the copy very cheap.

We tried, but ultimately we didn't. It was merged to master (31909515b37a096e62cbc23b63e4e3e4e574bfe7), then promptly reverted (5f4bf183874a3ff6492d85d27d5371f8cc428678), then there was another attempt (https://github.com/scylladb/scylladb/pull/8242), but it lost a merge race with `compound_sstable_set`, and it seems the project was abandoned at that point.

Reply:
@wmitros Why was versioned sstable set eventually abandoned?

Reply:
IIRC, the main reason for versioned sstable set was to address stalls, the stalls were (partially?) fixed, and we abandoned it due to cost benefit not being satisfactory. we can revisit it, and see if we can still benefit from it. with tablets, the value is even lower?!

Reply:
In v3 I kept the clone-modify-set logic.
We can consider something else as a follow-up